From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 0utplay <aldin@sijamhodzic.de>
Date: Thu, 5 Sep 2024 17:48:02 +0200
Subject: [PATCH] disable static argument suggestion validation


diff --git a/cloud-core/src/main/java/org/incendo/cloud/CommandTree.java b/cloud-core/src/main/java/org/incendo/cloud/CommandTree.java
index c18f80845f7bc2d756d195f7906681667b5a0831..781d615e4f1e0a2513704d26c9cb8d67eafdd619 100644
--- a/cloud-core/src/main/java/org/incendo/cloud/CommandTree.java
+++ b/cloud-core/src/main/java/org/incendo/cloud/CommandTree.java
@@ -756,7 +756,7 @@ public final class CommandTree<C> {
                     final String string = input.peekString();
                     for (Suggestion suggestion : suggestionsToAdd) {
                         if (suggestion.suggestion().equals(string) || !suggestion.suggestion().startsWith(string)) {
-                            continue;
+                            // continue; // CloudNet - disable suggestion pre-validation
                         }
                         context.addSuggestion(suggestion);
                     }
