From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derklaro <git@derklaro.dev>
Date: Sat, 15 Oct 2022 14:44:03 +0200
Subject: [PATCH] disable static argument suggestion validation


diff --git a/cloud-core/src/main/java/cloud/commandframework/CommandTree.java b/cloud-core/src/main/java/cloud/commandframework/CommandTree.java
index f94ba8440e0ede4229263589f83e2f26d6afff08..668239c78e25c999c6fbc9c32251b794e09cd80e 100644
--- a/cloud-core/src/main/java/cloud/commandframework/CommandTree.java
+++ b/cloud-core/src/main/java/cloud/commandframework/CommandTree.java
@@ -600,7 +600,7 @@ public final class CommandTree<C> {
                         .apply(commandContext, literalValue);
                 for (String suggestion : suggestionsToAdd) {
                     if (suggestion.equals(literalValue) || !suggestion.startsWith(literalValue)) {
-                        continue;
+                        // continue; // CloudNet - disable suggestion pre-validation
                     }
                     suggestions.add(suggestion);
                 }
