From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derklaro <git@derklaro.dev>
Date: Sat, 15 Oct 2022 14:44:03 +0200
Subject: [PATCH] disable static argument suggestion validation


diff --git a/cloud-core/src/main/java/cloud/commandframework/CommandTree.java b/cloud-core/src/main/java/cloud/commandframework/CommandTree.java
index 9f507c88a41940a366e7e7dc37185039f9302dc5..e4f114fe7082d8c8da857edecb092fd2afe4419f 100644
--- a/cloud-core/src/main/java/cloud/commandframework/CommandTree.java
+++ b/cloud-core/src/main/java/cloud/commandframework/CommandTree.java
@@ -602,7 +602,7 @@ public final class CommandTree<C> {
                         .apply(commandContext, literalValue);
                 for (String suggestion : suggestionsToAdd) {
                     if (suggestion.equals(literalValue) || !suggestion.startsWith(literalValue)) {
-                        continue;
+                        // continue; // CloudNet - disable suggestion pre-validation
                     }
                     suggestions.add(suggestion);
                 }
